The classical definition of heterozygosity describes heterozygosity as a measure of allelic divergence, where alleles are defined through chromosome pairing [56]. This definition poses a problem for genomes where chromosome pairing is not known (e.g., in polyploid genomes), as well as for genomes of mitotic asexuals where chromosome pairing may not occur at all. In these cases, concepts such as divergence between alleles (i.e., heterozygosity) vs. divergence of paralogs, that are clearly distinct in diploids, become blurred. We therefore quantify heterozygosity as the proportion of nucleotides that differ in at least one of the homologous chromosomes. Three examples of genomes with similar heterozygosity but different ploidy levels are shown on the scheme for illustration. Grey bars highlight specific loci, coloured bars represent alternative alleles at a given locus. 